File: iOS/ssabree/ssabree/Features/GroupDetail/GroupDetailViewModel.swift
import Foundation
import Observation

struct GroupDetailUiState {
    var detail: GroupDetailModel? = nil
    var members: [GroupMemberModel] = []
    var isLoading: Bool = false
    var error: String? = nil
    var isLeader: Bool = false
    var isMember: Bool = false
    var hasApplied: Bool = false  // 이미 지원했는지 여부
}

@Observable
final class GroupDetailViewModel {
    private let groupRepository: GroupRepository
    private let authRepository: AuthRepository
    private let groupKind: GroupKind
    private let groupId: Int

    var uiState = GroupDetailUiState()

    init(
        groupRepository: GroupRepository,
        authRepository: AuthRepository,
        groupKind: GroupKind,
        groupId: Int
    ) {
        self.groupRepository = groupRepository
        self.authRepository = authRepository
        self.groupKind = groupKind
        self.groupId = groupId
    }

    @MainActor
    func loadDetail() async {
        uiState.isLoading = true
        uiState.error = nil

        // Load group detail based on kind
        let detailResult: Result<GroupDetailModel, Error>
        if groupKind == .study {
            detailResult = await groupRepository.getStudyDetail(studyId: groupId)
        } else {
            detailResult = await groupRepository.getTeamDetail(teamId: groupId)
        }

        // Get current user ID
        let currentUserId = try? await authRepository.getMyMemberId().get()

        // Check if user is a member
        let myGroupsResult: Result<[GroupSummaryModel], Error>
        if groupKind == .study {
            myGroupsResult = await groupRepository.getMyStudies()
        } else {
            myGroupsResult = await groupRepository.getMyTeams()
        }
        let myGroups = (try? myGroupsResult.get()) ?? []
        let isMember = myGroups.contains { $0.id == groupId }

        // Check if user has already applied
        let myApplicationsResult: Result<[MyApplicationModel], Error>
        if groupKind == .study {
            myApplicationsResult = await groupRepository.getMyStudyApplications()
        } else {
            myApplicationsResult = await groupRepository.getMyTeamApplications()
        }
        let myApplications = (try? myApplicationsResult.get()) ?? []
        let hasApplied = myApplications.contains { $0.groupId == groupId && $0.status == "PENDING" }

        switch detailResult {
        case .success(let detail):
            let isLeader = currentUserId != nil && detail.leaderId == currentUserId
            uiState.detail = detail
            uiState.isLeader = isLeader
            uiState.isMember = isMember
            uiState.hasApplied = hasApplied
            uiState.isLoading = false

            // Load members
            await loadMembers()

        case .failure(let error):
            uiState.error = error.localizedDescription
            uiState.isLoading = false
        }
    }

    @MainActor
    private func loadMembers() async {
        let membersResult: Result<[GroupMemberModel], Error>
        if groupKind == .study {
            membersResult = await groupRepository.getStudyMembers(studyId: groupId)
        } else {
            membersResult = await groupRepository.getTeamMembers(teamId: groupId)
        }

        if case .success(let members) = membersResult {
            uiState.members = members
        }
    }

    func categoryLabel(for type: String) -> String {
        if groupKind == .study {
            return GroupTypeMapper.studyTypeToLabel(type)
        } else {
            return GroupTypeMapper.teamTypeToLabel(type)
        }
    }

    @MainActor
    func deleteGroup() async {
        uiState.isLoading = true
        uiState.error = nil

        let result: Result<Void, Error>
        if groupKind == .study {
            result = await groupRepository.deleteStudy(studyId: groupId)
        } else {
            result = await groupRepository.deleteTeam(teamId: groupId)
        }

        switch result {
        case .success:
            uiState.isLoading = false
        case .failure(let error):
            uiState.error = error.localizedDescription
            uiState.isLoading = false
        }
    }

    @MainActor
    func clearError() {
        uiState.error = nil
    }
}

File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/groupdetail/screen/GroupDetailViewModel.kt
package com.ssafy.ssabree.features.groupdetail.screen

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.repository.GroupRepository
import com.ssafy.ssabree.features.group.model.GroupDetailUiModel
import com.ssafy.ssabree.features.group.model.GroupKind
import com.ssafy.ssabree.features.group.model.toUiModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch

data class GroupDetailUiState(
    val detail: GroupDetailUiModel? = null,
    val isLoading: Boolean = false,
    val isLeader: Boolean = false,
    val isMember: Boolean = false,
    val hasPendingApplication: Boolean = false,
    val errorMessage: String? = null
)

class GroupDetailViewModel(
    private val groupRepository: GroupRepository,
    private val memberRepository: com.ssafy.ssabree.core.repository.MemberRepository,
    private val groupKind: GroupKind,
    private val groupId: Long
) : ViewModel() {

    private val _uiState = MutableStateFlow(GroupDetailUiState())
    val uiState: StateFlow<GroupDetailUiState> = _uiState.asStateFlow()

    init {
        loadDetail()
    }

    fun loadDetail() = viewModelScope.launch {
        _uiState.update { it.copy(isLoading = true, errorMessage = null) }
        val result = if (groupKind == GroupKind.STUDY) {
            groupRepository.getStudyDetail(groupId)
        } else {
            groupRepository.getTeamDetail(groupId)
        }
        val membersResult = if (groupKind == GroupKind.STUDY) {
            groupRepository.getStudyMembers(groupId)
        } else {
            groupRepository.getTeamMembers(groupId)
        }
        val myId = memberRepository.getMyMemberId().getOrNull()
        val myGroups = if (groupKind == GroupKind.STUDY) {
            groupRepository.getMyStudies().getOrNull()
        } else {
            groupRepository.getMyTeams().getOrNull()
        } ?: emptyList()
        val myApplications = if (groupKind == GroupKind.STUDY) {
            groupRepository.getMyStudyApplications().getOrNull()
        } else {
            groupRepository.getMyTeamApplications().getOrNull()
        } ?: emptyList()

        result.onSuccess { detail ->
            val currentMembers = membersResult.getOrNull()?.size ?: detail.currentMembers
            val mergedDetail = if (currentMembers != null) detail.copy(currentMembers = currentMembers) else detail
            val isLeader = myId != null && detail.leaderId != null && detail.leaderId == myId
            val isMember = myGroups.any { it.id == groupId }
            val hasPendingApplication = myApplications.any { app ->
                app.groupId == groupId && app.status == "PENDING"
            }
            _uiState.update {
                it.copy(
                    detail = mergedDetail.toUiModel(groupKind),
                    isLoading = false,
                    isLeader = isLeader,
                    isMember = isMember,
                    hasPendingApplication = hasPendingApplication
                )
            }
        }.onFailure { e ->
            _uiState.update { it.copy(isLoading = false, errorMessage = e.message) }
        }
    }

    fun deleteGroup(onSuccess: () -> Unit) = viewModelScope.launch {
        val result = if (groupKind == GroupKind.STUDY) {
            groupRepository.deleteStudy(groupId)
        } else {
            groupRepository.deleteTeam(groupId)
        }
        result.onSuccess { onSuccess() }
            .onFailure { e -> _uiState.update { it.copy(errorMessage = e.message) } }
    }

    fun clearError() {
        _uiState.update { it.copy(errorMessage = null) }
    }
}
